Subject:  Allow to use DeepCC on real networks.

Similar to https://github.com/Soheil-ab/Orca/issues/1#issuecomment-683734454

This file has inconsistent line endings.

LF files are real-network.sh, run-dcubic.sh,
CRLF files are server.cc

diff -urpN deepcc-1.0_p9999.orig/deepcc.v1.0/real-network.sh deepcc-1.0_p9999/deepcc.v1.0/real-network.sh
--- deepcc-1.0_p9999.orig/deepcc.v1.0/real-network.sh	1969-12-31 16:00:00.000000000 -0800
+++ deepcc-1.0_p9999/deepcc.v1.0/real-network.sh	2023-02-13 03:23:23.198694382 -0800
@@ -0,0 +1,26 @@
+if [ $# -ne 0 ]
+then
+    echo
+    echo "usage: $0"
+    echo
+    echo "Environment variables:"
+    echo
+    echo "DEEPCC_SCHEME - cubic, westwood, illi, bbr, ..."
+    echo
+    echo "The prebuilt learned models are listed, but different models require"
+    echo "training."
+    echo
+    exit
+fi
+sudo sysctl -w -q net.ipv4.ip_forward=1
+sudo sysctl -w -q net.ipv4.tcp_no_metrics_save=1
+port=8000
+scheme=${DEEPCC_SCHEME:-cubic}
+
+trace_index=25      # Check setup.sh for the valid trace indices.
+
+latency=10 # Unidirectional Link Delay ==> minRTT>=2*latency (ms)
+target=50  # ==>Overall application's Target delay. (ms)
+report_period=20
+./run-deep.sh $port ${trace_index} $target $report_period $latency 5 $scheme
+
diff -urpN deepcc-1.0_p9999.orig/deepcc.v1.0/run-dcubic.sh deepcc-1.0_p9999/deepcc.v1.0/run-dcubic.sh
--- deepcc-1.0_p9999.orig/deepcc.v1.0/run-dcubic.sh	2023-02-13 03:05:10.662658380 -0800
+++ deepcc-1.0_p9999/deepcc.v1.0/run-dcubic.sh	2023-02-13 03:20:19.242688320 -0800
@@ -57,7 +57,7 @@ sudo killall rl-server mm-link client
 sudo killall -s15 python
 sleep 10
 echo "Experiment is finished."
-if [ ${training} -eq 2 ] || [ ${training} -eq 4 ]
+if [ ${training} -eq 2 ] || [ ${training} -eq 4 ] || [ ${training} -eq 5 ]
 then
     echo "Doing analysis ..."
     out="$prefix-${log}.tr"
diff -urpN deepcc-1.0_p9999.orig/deepcc.v1.0/src/server.cc deepcc-1.0_p9999/deepcc.v1.0/src/server.cc
--- deepcc-1.0_p9999.orig/deepcc.v1.0/src/server.cc	2023-02-13 03:05:10.666658381 -0800
+++ deepcc-1.0_p9999/deepcc.v1.0/src/server.cc	2023-02-13 03:28:22.910704258 -0800
@@ -78,7 +78,7 @@ int main(int argc, char **argv)
 
 void usage()
 {
-	DBGMARK(0,0,"./server [Delay(ms)] [port] [DL-trace] [UP-trace] [log] [Target] [Initial Alpha] [qsize in pkts] [Report Period: 1 sec] [AQM:2 FIFO] [First Time: 1=yes, 0=no] [tcp:cubic westwood illinois bbr ...]\n");
+	DBGMARK(0,0,"./server [Delay(ms)] [port] [DL-trace] [UP-trace] [log] [Target] [Initial Alpha] [qsize in pkts] [Report Period: 1 sec] [AQM:2 FIFO] [First Time: 1=yes, 0=no, 2=evaluate, 4=test, 5=real-network(production)] [tcp:cubic westwood illinois bbr ...]\n");
 }
 
 void start_server(int flow_num, int client_port)
@@ -149,15 +149,21 @@ void start_server(int flow_num, int clie
         }
     }
     char container_cmd[500];
-    sprintf(container_cmd,"sudo ./client $MAHIMAHI_BASE 1 %d",client_port);
-    for(int i=1;i<FLOW_NUM;i++)
+    if (first_time!=5)
     {
-    	sprintf(container_cmd,"%s & sleep 5;sudo ./client $MAHIMAHI_BASE 1 %d",container_cmd,client_port+i);
+        sprintf(container_cmd,"sudo ./client $MAHIMAHI_BASE 1 %d",client_port);
+        for(int i=1;i<FLOW_NUM;i++)
+        {
+    	    sprintf(container_cmd,"%s & sleep 5;sudo ./client $MAHIMAHI_BASE 1 %d",container_cmd,client_port+i);
+        }
     }
     char cmd[1000];
     char final_cmd[1000];
-    sprintf(cmd, "sudo -u `logname` mm-delay %d mm-link ../traces/%s ../traces/%s --downlink-log=log/down-%s --uplink-queue=droptail --uplink-queue-args=\"packets=%d\" --downlink-queue=droptail --downlink-queue-args=\"packets=%d\" -- sh -c \'%s\' &",delay_ms,uplink,downlink,log_file,qsize,qsize,container_cmd);   
-    sprintf(final_cmd,"%s",cmd);
+    if (first_time!=5)
+    {
+        sprintf(cmd, "sudo -u `logname` mm-delay %d mm-link ../traces/%s ../traces/%s --downlink-log=log/down-%s --uplink-queue=droptail --uplink-queue-args=\"packets=%d\" --downlink-queue=droptail --downlink-queue-args=\"packets=%d\" -- sh -c \'%s\' &",delay_ms,uplink,downlink,log_file,qsize,qsize,container_cmd);   
+        sprintf(final_cmd,"%s",cmd);
+    }
     info->trace=trace;
     info->num_lines=num_lines;
     /**
@@ -195,7 +201,7 @@ void start_server(int flow_num, int clie
                 (double)target,congestion,(int)key,(int)key_rl);
         DBGPRINT(0,0,"Starting RL Module (Without load) ...\n");
     }
-    else if (first_time==2 || first_time==4){
+    else if (first_time==2 || first_time==4 || first_time==5){
         sprintf(cmd,"sudo /home/`logname`/venv/bin/python drl_agent.py --target=%.7f --tb_interval=1 --load=1 --eval --scheme=%s --train_dir=\".\" --mem_r=%d --mem_w=%d &",
                 (double)target,congestion,(int)key,(int)key_rl);
         DBGPRINT(0,0,"Starting RL Module (No learning) ...\n");
@@ -297,7 +303,11 @@ void* TimerThread(void* information)
     target_array[1]=target;
     target_array[2]=(target*3)/2;
 #ifdef CHANGE_TARGET
-    if (first_time!=2)
+    if (first_time==2)
+        ;
+    else if (first_time==5)
+        ;
+    else
     {
         while(true)
         {
